bigframes/core/__init__.py

```diff
@@ -453,6 +453,19 @@ def join(
             return ArrayValue(bigframes.core.rewrite.maybe_rewrite_join(join_node))
         return ArrayValue(join_node)
 
+    def try_align_as_projection(
+        self,
+        other: ArrayValue,
+        join_type: join_def.JoinType,
+        mappings: typing.Tuple[join_def.JoinColumnMapping, ...],
+    ) -> typing.Optional[ArrayValue]:
```

Comment:
Aside: This is slightly safer to return None now, since we have type checking. But, it's not universally applied (see all the `note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]` warnings in our mypy session.

I created Draft PR https://github.com/googleapis/python-bigquery-dataframes/pull/693 to fix that, but there's `151 errors in 32 files (checked 187 source files)` to cleanup.

Reply:
I tend to prefer capturing expected failures in the return value rather than as an exception, for the type safety, and easier composition. The calling code can provide a more informative exception anyways and errors like this don't need a stack trace.

Reply:
Resolving. I feel better about this now that https://github.com/googleapis/python-bigquery-dataframes/pull/693 is merged.